builtins/web/form-data/form-data-encoder.cpp

```diff
@@ -0,0 +1,492 @@
+#include "form-data-encoder.h"
+#include "extension-api.h"
+#include "form-data.h"
+
+#include "../base64.h"
+#include "../blob.h"
+#include "../file.h"
+#include "../streams/buf-reader.h"
+#include "../streams/native-stream-source.h"
+
+#include "encode.h"
+#include "mozilla/Assertions.h"
+
+#include <fmt/format.h>
+#include <string>
+
+namespace {
+
+const char LF = '\n';
+const char CR = '\r';
+const char *CRLF = "\r\n";
+
+size_t compute_normalized_len(std::string_view src, const char *newline) {
+  size_t len = 0;
+  size_t newline_len = strlen(newline);
+
+  for (size_t i = 0; i < src.size(); i++) {
+    if (src[i] == CR) {
+      if (i + 1 < src.size() && src[i + 1] == LF) {
+        len += newline_len;
+        i++;
+      } else {
+        len += newline_len;
+      }
+    } else if (src[i] == LF) {
+      len += newline_len;
+    } else {
+      len += 1;
+    }
+  }
+
+  return len;
+}
+
+// Replace every occurrence of U+000D (CR) not followed by U+000A (LF),
+// and every occurrence of U+000A (LF) not preceded by U+000D (CR),
+// in entry's name, by a string consisting of a U+000D (CR) and U+000A (LF).
+//
+// https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#multipart-form-data
+std::optional<std::string> normalize_newlines(std::string_view src) {
+  std::string output;
+
+  output.reserve(compute_normalized_len(src, CRLF));
+
+  for (size_t i = 0; i < src.size(); i++) {
+    if (src[i] == CR) {
+      if (i + 1 < src.size() && src[i + 1] == LF) {
+        output += CRLF;
+        i++;
+      } else {
+        output += CRLF;
+      }
+    } else if (src[i] == LF) {
+      output += CRLF;
+    } else {
+      output.push_back(src[i]);
+    }
+  }
+
+  return output;
+}
+
+std::optional<std::string> normalize_newlines(JSContext *cx, HandleValue src) {
+  auto chars = core::encode(cx, src);
+  if (!chars) {
+    return std::nullopt;
+  }
+
+  return normalize_newlines(chars);
+}
+
+// For field names and filenames for file fields, the result of the encoding must
+// be escaped by replacing any 0x0A (LF) bytes with the byte sequence
+// `%0A`, 0x0D (CR) with `%0D` and 0x22 (") with `%22`.
+//
+// https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#multipart-form-data
+std::optional<std::string> escape_newlines(std::string_view str) {
```

Comment:
Since this isn't just for newlines, it'd be good to give it a different name. Perhaps `escape_name`, given that this applies escaping of a set of characters in names?

Reply:
I've changed this to `escape_name`.